[#10823] student accessing an awaiting session: error message not shown

lyeyixian opened the pull request:
Fixes #10823  

**Outline of Solution**

I found out that this bug happened when the instructor changed the opening time of a session to a later date, making it not visible to students anymore. But, using the same link, the students are still able to go to the page.

The status code is 403, but in the frontend code, only 404 is checked. So, I added a check for 403 and update the error message.

**Old**

![image](https://user-images.githubusercontent.com/58477697/125773333-08e2638b-d031-4b34-ada1-025dad794f52.png)


**Updated**

![image](https://user-images.githubusercontent.com/58477697/125773239-b2c79527-22ca-449a-b74b-49cfd988655c.png)

halfwhole reviewed (changes requested):
The changes look good! We should also update the session submission page tests to test for this particular error as well

jianhandev commented on src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:
```diff
@@ -341,8 +341,14 @@ this session.`;
         }
+
+        if (resp.status === 403) {
+          this.simpleModalService.openInformationModal('Feedback Session Is Not Visible!', SimpleModalType.DANGER,
+            'The session is not visible (most likely updated by the instructor after the submission link was sent).');
+        }
```
About the error message, personally I feel that `Feedback session is not yet open` is more intuitive to the user. Perhaps can also hint at the user to access the session again once the session opens.

lyeyixian replied:
OKOK, I will update it.

jianhandev reviewed (commented):
Changes look good, however I feel that the error message to the user can be improved (see comment below).

lyeyixian commented:
> The changes look good! We should also update the session submission page tests to test for this particular error as well

Yup, already update the test.

halfwhole reviewed (approved):
LGTM for the code!

Just to confirm, @damithc how do these new UI changes look?

<img width="1124" alt="Screenshot 2021-07-27 at 10 08 37 PM" src="https://user-images.githubusercontent.com/41856541/127168623-1fb46ac0-c7a5-4db5-83e0-c8e790446ec6.png">

damithc commented:
> Just to confirm, @damithc how do these new UI changes look?
> 
> <img alt="Screenshot 2021-07-27 at 10 08 37 PM" width="1124" src="https://user-images.githubusercontent.com/41856541/127168623-1fb46ac0-c7a5-4db5-83e0-c8e790446ec6.png">

The dialog box is fine. There shouldn't be a candy bar (seen at the top) as well right?

rrtheonlyone commented:
> There shouldn't be a candy bar (seen at the top) as well right?

@damithc would you prefer the toast at the top or you rather have the info modal? From an implementation perspective, it is easier to show the toast with the same message.

damithc commented:
> @damithc would you prefer the toast at the top or you rather have the info modal? From an implementation perspective, it is easier to show the toast with the same message.

Given the message is long and intended for the user to read fully, I think a modal is more appropriate.

rrtheonlyone commented:
Are we okay to leave the toast message at the top @damithc along with the info modal.

damithc commented:
> Are we okay to leave the toast message at the top @damithc along with the info modal?

🤔 I think that's a bit odd. The toast message is redundant. Is it a lot of extra work to omit it?

rrtheonlyone reviewed (approved):
No realised it is an easy change - @lyeyixian to change the below and we are good to go!

Update `this.navigationService.navigateWithErrorMessage(this.router, '/web/student/home', resp.error.message);` to `this.navigationService.navigateByURL(this.router, '/web/student/home');` This is in line 351.

lyeyixian commented:
> No realised it is an easy change - @lyeyixian please change the below and we are good to go!

Already changed it. Thanks for the help!

wkurniawan07 commented on src/web/app/pages-session/session-submission-page/session-submission-page.component.ts:
```diff
@@ -341,8 +341,14 @@ this session.`;
+
+        if (resp.status === 403) {
+          this.simpleModalService.openInformationModal('Feedback Session Is Not Yet Open!', SimpleModalType.DANGER,
+            'The session is not yet open (the submission opening time is most likely updated by the instructor after the submission link was sent). <br/>You can access the session again once the session opens.');
```
Is 403 _always_ equivalent to session not open? For all we know it could be an open session legitimately not visible to the user.

lyeyixian replied:
Oh ya, there could be other cases also. Hmm, I think we need to change the messages.

lyeyixian commented:
I updated the error message to make it more generic, but this makes it not that detailed for the users as compared to the previous one. 

I also added an else block so that everything is handled.

wkurniawan07 commented:
> I updated the error message to make it more generic, but this makes it not that detailed for the users as compared to the previous one.

How "generic" are we talking about here? The message looks like it is taken from `resp.error.message`, which may (or may not) be specific?

lyeyixian commented:
> How "generic" are we talking about here? The message looks like it is taken from `resp.error.message`, which may (or may not) be specific?

Sorry, yes I mean to say the message follows the caught error message, so it is not fixed to just "Feedback Session Not Visible" and can inform the user the reason for the error in a better way.

wkurniawan07 reviewed (approved):
LGTM
